feat: Refactor/Renew smart contracts

mscheong01 opened the pull request.

Commit:
feat: Waffle.sol core 로직 재정립

Commit:
fix: renew waffle minting function

Commit:
feat: added initial 10 token generation

Commit:
fix: compile errors

cs71107 commented on contracts/Waffle.sol:
```diff
@@ -7,28 +7,27 @@ import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+        verticals: 세로 줄 위치
         === verticals & horizontals
+        7*7 격자 상에 서로 수직인 평행선 두세트
         - msb => 1~5 (선 긋는 위치)
-         가로 * 세로 선 긋는 조합 = 100 가지 ( 1~5 중 2개, 5C2 = 10 )
+         가로 * 세로 선 긋는 조합 = 36가지
         - lsb => 색상 16개 (x0 ~xF)
```
36가지면 충분할까요? 약간 더 늘려도 상관 없을 것 같아서요

cs71107 commented on contracts/Waffle.sol:
```diff
@@ -42,34 +41,69 @@ contract Waffle is Ownable {
+        string memory name,
+        uint8[2] memory horizontals,
+        uint8[2] memory verticals
+    ) internal returns (uint) {
+        require(!_validateShape(horizontals, verticals), 'this shape is not a waffle at all!');
+
+        Flavor flavor = _getRandomFlavour(name);
+        waffles.push(MetaData(name, flavor, horizontals, verticals));
```
이 부분에 한 계정이 너무 많은 shape를 가지지 못하게 하는게 추가 되면 좋을 것 같습니다. 지금 36개 그대로 간다면 한 사람이 독점하기 너무 쉬울 것 같아요

mscheong01 replied:
이건 claim 가격을 올릴까 생각했었습니다

Jhvictor4 replied:
이거는 민팅 external 함수를 막아버리는 쪽으로 가면 될듯 하네요

cs71107 commented on contracts/Waffle.sol:
```diff
@@ -42,34 +41,69 @@ contract Waffle is Ownable {
 
-    function _getRandomLine(string memory title) internal returns(uint8) {
-        return uint8(_getRandom(title)) % 5 + 1;
+    function _validateShape(
+        uint8[2] memory horizontals,
+        uint8[2] memory verticals
+    ) internal returns (bool) {
+        if(verticals[0] >= verticals[1]) return false;
+        if(horizontals[0] >= horizontals[1]) return false;
+        uint8 vDistance = verticals[1] - verticals[0];
+        uint8 hDistance = horizontals[1] - horizontals[0];
+        if(vDistance == 1 || hDistance == 1) return false;
+        return true;
+    }
```
horizontal, verticals가 독립적으로 판단되는 것 같은데, 좀 더 쪼개도 되지 않을까요? (너무 과한가요..?)

mscheong01 replied:
더 쪼갠다는게 무슨말인지 설명 부탁드립니다

cs71107 replied:
그러니까 horizontals, verticals가 지금 true false를 판단할 때 서로 영향을 주지 않기 때문에, uint[2] memory 하나에 대해서 판단하는 함수를 짠 다음, 그걸 and 하는 식으로 쪼개도 괜찮지 않을까 생각했었어요. 좀 과한 것 같기도 하고요.

Jhvictor4 replied:
그렇게 합쳐서 얻는 이점이 있을까요?

cs71107 replied:
음... 저 같은 경우 확장이 좀 더 유리하고 나중에 수정도 용이할 거라고 생각하긴 했는데 생각해보니 지금은 굳이? 싶긴 하네요...

mscheong01 replied:
아 그거야 쪼갤수 있는거 맞고요, 커찮아서 요건 하나로 뒀습니다

cs71107 reviewed (commented):
전반적으로 좋은 것 같습니다. 약간 디테일하게 에러 메시지를 보여줄 수 있다면 더 좋을 것 같네요 (예를 들어 순서가 어디가 잘못 됐다 하는 식으로...)

Commit:
feat: added to validation

Commit:
feat: added token migration file

Commit:
fix: error

Commit:
Update Waffle.sol

Jhvictor4 committed:
Update .gitignore

Jhvictor4 committed:
deploy setting done

1. truffle migrate
2. truffle run verify WaffleToken --network ropsten

fixme 값들 알아서 잘 세팅하기

Jhvictor4 committed:
최종 수정

- id 건너뛰면서 초기화 하는 로직 삭제
 - sequential array를 자료구조로 사용하므로, 프로퍼티를 알맞게 가져올 수가 없음
 - 나중에 pseudo-random 사용하는 게 맞아 보임
- show Horizontals, Verticals 추가
- _baseURI override
- state mutability 추가

- 와플 자료구조 mapping으로 변경 (맘에 안듬..)

- 테스트넷 배포 시 컨펌 0회로 수정
